File: findMinimum.java
// Time Complexity O(logn)
// Space Complexity O(1)
class Solution {
    public int findMin(int[] nums) {
        int start = 0;
        int end = nums.length - 1;
        
        while(start <= end) {
            if(nums[start] <= nums[end]) return nums[start];
            
            int mid = start + (end - start) / 2;
            
            boolean leftgt = mid == 0 || nums[mid] > nums[mid-1];
            boolean rightgt = mid == nums.length - 1 || nums[mid] > nums[mid+1];
            
            if(leftgt && rightgt) {
                return nums[mid];
            }
            
            if(leftgt) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
            
            
        }
        
        return -1;
    }
}

File: findMinimum.py
# Ran on Leetcode - No



class Solution:
    def findMin(self, nums: List[int]) -> int:
        
        left = 0
        right = len(nums) - 1
        
        while left <= right:
            
            mid = left + (right-left)//2
            
            if nums[mid] >= nums[left] and nums[mid] <= nums[right]:
                
                return nums[left]
            
            if nums[mid] >= nums[left]:
                # left is sorted
                left = mid 
                
            elif nums[mid] <= nums[right]:
                right = mid 
            